src/typechecker/base.lisp

```diff
@@ -14,25 +14,74 @@
    #:make-coalton-error-help)
   (:local-nicknames
    (#:util #:coalton-impl/util)
-   (#:error #:coalton-impl/error)
-   (#:tc #:coalton-impl/typechecker/stage-1))
+   (#:error #:coalton-impl/error))
   (:export
+   #:*coalton-pretty-print-tyvars*
+   #:*pprint-tyvar-dict*
+   #:*pprint-variable-symbol-code*
+   #:*pprint-variable-symbol-suffix*
    #:tc-error                           ; CONDITION
    #:check-duplicates                   ; FUNCTION
    #:check-package                      ; FUNCTION
+   #:with-pprint-variable-scope         ; MACRO
+   #:with-pprint-variable-context       ; MACRO
    ))
 
 (in-package #:coalton-impl/typechecker/base)
 
+;;;
+;;; Shared definitions for type checking environment
+;;;
+
+;;;
+;;; Pretty printing
+;;;
+
+(defvar *pprint-variable-symbol-code*)
+
+(defvar *pprint-variable-symbol-suffix*)
+
+(defmacro with-pprint-variable-scope (() &body body)
```

Comment:
**not** for this PR, but this casually duplicates code, which could lead to unintentional code bloat.

solutions:

- bind variables to a sentinel and slide the IF into the binding: `(let ((*x* (if (eq sentinel *x*) default *x* ...`
- put code into a DYNAMIC-EXTENT FLET: `(flet ((body () ...)) (declare (dx #'body)) (if ... (let ... (body)) (let ... (body))))`